bokehjs/src/coffee/models/tiles/mercator_tile_source.coffee

```diff
@@ -111,12 +112,20 @@ export class MercatorTileSource extends TileSource
       return previous
     return @_resolutions.indexOf(closest)
 
-  snap_to_zoom: (extent, height, width, level) ->
+  snap_to_zoom_level: (extent, height, width, level) ->
+    [xmin, ymin, xmax, ymax] = extent
     desired_res = @_resolutions[level]
     desired_x_delta = width * desired_res
     desired_y_delta = height * desired_res
-
-    [xmin, ymin, xmax, ymax] = extent
+    if !@snap_to_zoom
+      xscale = (xmax-xmin)/desired_x_delta
+      yscale = (ymax-ymin)/desired_y_delta
+      if xscale > yscale
+        desired_x_delta = (xmax-xmin)
+        desired_y_delta = desired_y_delta*xscale
+      else
+        desired_x_delta = desired_x_delta*yscale
+        desired_y_delta = (ymax-ymin)
```

Comment:
There may be re-usable code around `match_aspect` code paths. In any case anything that is a simple mathematical computation would be good to factor out since it can then be tested with simple unit tests.

Reply:
I had a look and couldn't find anything directly relevant. I think the problem is a bit better constrained than in the ``match_aspect`` case anyway.